File: c++/leetcode/543. Diameter of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
  int diameter(TreeNode *, int &);
  int diameterOfBinaryTree(TreeNode *root)
  {
    int ans = 0;
    diameter(root, ans);
    return ans;
  }
};

int Solution::diameter(TreeNode *root, int &ans)
{
  if (!root)
    return 0;

  int left = diameter(root->left, ans);
  int right = diameter(root->right, ans);

  ans = max(ans, left + right);
  return max(left, right) + 1;
}

File: javascript/leetcode/543. Diameter of Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function (root) {
  let ans = 0;
  let dfs = (function dim(root) {
    if (!root) return 0;

    const left = dim(root.left);
    const right = dim(root.right);

    ans = Math.max(ans, left + right);

    return Math.max(left, right) + 1;
  })(root);

  return ans;
};
